posts/2021-04-21-What I wish somebody told me when I was learning Haskell.md

```diff
@@ -0,0 +1,337 @@
+
+## Embrace purity
+
+### Functions do nothing
+
+In a functional language a function does nothing - it simply defines a transformation between types - but it does not perform this transformation until it is... It is tempting to say "called", but the functions are not "called" in functional programming languages in the same way - they are evaluated and combined in bigger transformations, until the whole program is constructed.
+
+Only when the whole program is executed something happens - but you cannot simply break down program execution to the individual function calls - compiler can produce a more efficient machine code than if functions were individually called.
+
+Every function definition can be replaced with its code - there is no such thing as impure function in Haskell (and yes, that includes functions that operate on Monads, whether they use `do` notation or not, - they are also pure functions).
+
+### Function evaluation produces no side effects
+
+This is the consequence of the fact that the function does nothing. To have a side effect it would have to do something, but it does not. To my and many people surprise, this statement holds true even for functions that evaluate to the type of IO monad - it is conventional to say that such functions perform IO, are executed step by step, and have side effects.
+
+But in reality it is not function evaluation that produces side effects - a function simply evaluates to IO action. This action can be chained with another IO action or it can be passed as parameter and simply be thrown away based on some other function evaluation. The whole program is one combined IO action that is, when executed, does indeed produce side effects.
+
+Many tutorials separate code to functions with and without side effects, but it only increases the confusion about how the language works, and confronted with a simple function of the type `Bool -> IO a -> IO b -> IO (Either a b)` most beginners (myself included) would be puzzled by what this function can possibly do.
+
+But once you remind yourself that functions do nothing, you may realize that this function probably evaluates to a single IO action by choosing one of two provided IO actions based on the boolean value (and none of the actions is performed, until the whole program runs):
+
+```haskell
+ifIO: Bool -> IO a -> IO b -> IO (Either a b)
+ifIO x a b = if x then Right <$> b else Left <$> a
+```
+
+## Choose tools and libraries
+
+### Effective developer environment for fast iteration
+
+My choices, mostly suggested by some very pragmatic people, were:
+
+- stack, to avoid managing ghc and cabal separately, and always have compatible dependency versions that were tested against each other on stackage.org
+- simple-hpack template, for better dependency management
+- ormolu for code formatting
+- hlint - it is very helpful with its non-trivial suggestions on how the code could be simplified, making your code much more concise.
```

Comment:
Mention Haskell Language Server for IDE support? For example this extension that uses it works well now and includes both hlint and ormolu out of the box (with possibility for configuration IIRC)
https://marketplace.visualstudio.com/items?itemName=haskell.haskell